cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -0,0 +1,321 @@
+            sdistOutputPath (\o flags -> flags { sdistOutputPath = o })
+            (reqArg "PATH" (succeedReadE Flag) flagToList)
+        ]
+    }
+
+data SdistFlags = SdistFlags
+    { sdistVerbosity     :: Flag Verbosity
+    , sdistDistDir       :: Flag FilePath
+    , sdistProjectFile   :: Flag FilePath
+    , sdistListSources   :: Flag Bool
+    , sdistNulSeparated  :: Flag Bool
+    , sdistArchiveFormat :: Flag ArchiveFormat
+    , sdistOutputPath    :: Flag FilePath
+    }
+
+defaultSdistFlags :: SdistFlags
+defaultSdistFlags = SdistFlags
+    { sdistVerbosity     = toFlag normal
+    , sdistDistDir       = mempty
+    , sdistProjectFile   = mempty
+    , sdistListSources   = toFlag False
+    , sdistNulSeparated  = toFlag False
+    , sdistArchiveFormat = toFlag TargzFormat
+    , sdistOutputPath    = mempty
+    }
+
+--
+
+sdistAction :: SdistFlags -> [String] -> GlobalFlags -> IO ()
+sdistAction SdistFlags{..} targetStrings globalFlags = do
+    let verbosity = fromFlagOrDefault normal sdistVerbosity
+        mDistDirectory = flagToMaybe sdistDistDir
+        mProjectFile = flagToMaybe sdistProjectFile
+        globalConfig = globalConfigFile globalFlags
+        listSources = fromFlagOrDefault False sdistListSources
+        nulSeparated = fromFlagOrDefault False sdistNulSeparated
+        archiveFormat = fromFlagOrDefault TargzFormat sdistArchiveFormat
+        mOutputPath = flagToMaybe sdistOutputPath
+  
+    projectRoot <- either throwIO return =<< findProjectRoot Nothing mProjectFile
+    let distLayout = defaultDistDirLayout projectRoot mDistDirectory
+    dir <- getCurrentDirectory
+    projectConfig <- runRebuild dir $ readProjectConfig verbosity globalConfig distLayout
+    baseCtx <- establishProjectBaseContext verbosity projectConfig
+    let localPkgs = localPackages baseCtx
+
+    targetSelectors <- either (reportTargetSelectorProblems verbosity) return
+        =<< readTargetSelectors localPkgs targetStrings
+    
+    mOutputPath' <- case mOutputPath of
+        Just "-"  -> return (Just "-")
+        Just path -> Just <$> makeAbsolute path
+        Nothing   -> return Nothing
+    
+    let 
+        format =
+            if | listSources, nulSeparated -> SourceList '\0'
+               | listSources               -> SourceList '\n'
+               | otherwise                 -> Archive archiveFormat
+
+        ext = case format of
+                SourceList _        -> "list"
+                Archive TargzFormat -> "tar.gz"
+                Archive ZipFormat   -> "zip"
+    
+        outputPath pkg = case mOutputPath' of
+            Just path
+                | path == "-" -> "-"
+                | otherwise   -> path </> prettyShow (packageId pkg) <.> ext
+            Nothing
+                | listSources -> "-"
+                | otherwise   -> distSdistFile distLayout (packageId pkg) archiveFormat
+
+    createDirectoryIfMissing True (distSdistDirectory distLayout)
```

Comment:
Strictly speaking, this is not needed when outputting to stdout, but no-one probably cares.